src/components/control-bar/ClosedCaptionButton.js

```diff
@@ -0,0 +1,140 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import classNames from 'classnames';
+import MenuButton from '../menu/MenuButton';
+
+const propTypes = {
+  player: PropTypes.object,
+  actions: PropTypes.object,
+  className: PropTypes.string,
+  kinds: PropTypes.array
+};
+
+const defaultProps = {
+  kinds: ['captions', 'subtitles'] // `kind`s of TextTrack to look for to associate it with this menu.
+};
+
+class ClosedCaptionButton extends Component {
+  constructor(props, context) {
+    super(props, context);
+
+    this.getTextTrackItems = this.getTextTrackItems.bind(this);
+    this.updateState = this.updateState.bind(this);
+    this.handleSelectItem = this.handleSelectItem.bind(this);
+
+    this.state = this.getTextTrackItems();
+  }
+
+  componentDidUpdate() {
+    this.updateState();
+  }
+
+  getTextTrackItems() {
+    const { kinds, player } = this.props;
+    const { textTracks, activeTextTrack } = player;
+    const textTrackItems = {
+      items: [],
+      selectedIndex: 0
+    };
+    const tracks = Array.from(textTracks || []);
+
+    if (tracks.length === 0) {
+      return textTrackItems;
+    }
+
+    textTrackItems.items.push({
+      label: 'Off',
+      value: null
+    });
+
+    tracks.forEach((textTrack) => {
+      // ignore invalid text track kind
+      if (kinds.length && !kinds.includes(textTrack.kind)) {
+        return;
+      }
+
+      textTrackItems.items.push({
+        label: textTrack.label,
+        value: textTrack.language
+      });
+    });
+
+    textTrackItems.selectedIndex = textTrackItems.items.findIndex(
+      item => activeTextTrack && activeTextTrack.language === item.value
+    );
+
+    if (textTrackItems.selectedIndex === -1) {
+      textTrackItems.selectedIndex = 0;
+    }
+
+    return textTrackItems;
+  }
+
+  updateState() {
+    const textTrackItems = this.getTextTrackItems();
+    if (
+      textTrackItems.selectedIndex !== this.state.selectedIndex
+      || !this.textTrackItemsAreEqual(textTrackItems.items, this.state.items)
+    ) {
+      this.setState(textTrackItems);
+    }
+  }
+
+  textTrackItemsAreEqual(items1, items2) {
```

Comment:
I think we can move this function outside of the component

Reply:
This is the only place we need to compare two arrays, and it's only for textTrack items.